aten/src/ATen/tensor_options_utils.py

```diff
@@ -0,0 +1,181 @@
+
+        if has_TO:
+            index = [idx for idx, formal in enumerate(formals) if 'ScalarType dtype' in formal][0]
+            collapsed.pop(index)
+            collapsed.pop(index)
+            collapsed.pop(index)
+            collapsed.pop(index)
+            if has_def_val:
+                collapsed.insert(index, 'const at::TensorOptions & options={}')
+            else:
+                collapsed.insert(index, 'const at::TensorOptions & options')
+
+            return collapsed
+
+    return collapsed
+
+# Collapses tensor options formals into one TensorOptions formal argument object
+# with all the metadata.
+# Covers cases for optional and non-optional tensor options arguments
+def collapse_formals_list(formals):
+    collapsed = formals[:]
+    if (any(formal['type'] == 'c10::optional<ScalarType>' for formal in collapsed) and
+        any(formal['type'] == 'c10::optional<Layout>' for formal in collapsed) and
+        any(formal['type'] == 'c10::optional<Device>' for formal in collapsed) and
+            any(formal['type'] == 'c10::optional<bool>' for formal in collapsed)):
+        index = 0
+        for i in range(len(collapsed)):
+            if collapsed[i]['type'] == 'c10::optional<ScalarType>':
+                break
+            else:
+                index += 1
+
+        collapsed.pop(index)
+        collapsed.pop(index)
+        collapsed.pop(index)
+        collapsed.pop(index)
+        collapsed.insert(index, {"annotation" : "None",
+                                 "dynamic_type": "TensorOptions",
+                                 "is_nullable": "False",
+                                 "default": "{}",
+                                 "kwarg_only": "True",
+                                 "name": "options",
+                                 "type": "const TensorOptions &", })
+
+    if (any(formal['type'] == 'ScalarType' for formal in collapsed) and
+        any(formal['type'] == 'Layout' for formal in collapsed) and
+        any(formal['type'] == 'Device' for formal in collapsed) and
+            any(formal['type'] == 'bool' for formal in collapsed)):
+        index = 0
+        for i in range(len(collapsed)):
+            if collapsed[i]['type'] == 'ScalarType':
+                break
+            else:
+                index += 1
+
+        collapsed.pop(index)
+        collapsed.pop(index)
+        collapsed.pop(index)
+        collapsed.pop(index)
+        collapsed.insert(index, {"annotation" : "None",
+                                 "dynamic_type": "TensorOptions",
+                                 "is_nullable": "False",
+                                 "kwarg_only": "True",
+                                 "name": "options",
+                                 "type": "const TensorOptions &", })
+
+    return collapsed
+
+# This is a hack.
+# Please see [All schemas in native_functions.yaml that have TensorOptions
+# should be have optional ScalarType, Layout, Device and pin memory]
```

Comment:
The Note mechanism only works if you can actually easily grep for the Note; when the title is long enough that it breaks over a newline that is not good for grepping. Try to shorten the title.